randovania/games/dread/layout/dread_damage_state.py

```diff
@@ -0,0 +1,111 @@
+        self._use_immediate_energy_parts = use_immediate_energy_parts
+        self._energy_part_item = energy_part_item
+
+    def __copy__(self) -> Self:
+        return self._duplicate()
+
+    @override
+    def resources_for_energy(self) -> Generator[ItemResourceInfo]:
+        yield self._energy_tank
+        yield self._energy_part_item
+
+    def _maximum_energy(self, resources: ResourceCollection) -> int:
+        num_tanks = resources[self._energy_tank]
+        num_parts = resources[self._energy_part_item]
+        num_whole_tanks = num_tanks + num_parts // 4
+        num_partial_parts = num_parts % 4
+        energy_from_partial = (self._energy_per_tank * num_partial_parts) // 4
+
+        return self._starting_energy + (self._energy_per_tank * num_whole_tanks) + energy_from_partial
+
+    def _duplicate(self) -> Self:
+        result = DreadDamageState(
+            self._starting_energy,
+            self._energy_per_tank,
+            self._energy_tank,
+            self._use_immediate_energy_parts,
+            self._energy_part_item,
+        )
+        result._energy = self._energy
+        return result
+
+    def _energy_part_difference(
+        self,
+        new_resources: ResourceCollection,
+        old_resources: ResourceCollection,
+    ) -> int:
+        return new_resources[self._energy_part_item] - old_resources[self._energy_part_item]
+
+    @override
+    def apply_collected_resource_difference(
+        self, new_resources: ResourceCollection, old_resources: ResourceCollection
+    ) -> Self:
+        if self._energy_tank_difference(new_resources, old_resources) > 0:
+            return self._at_maximum_energy(new_resources)
+        elif self._energy_part_difference(new_resources, old_resources) > 0:
+            if self._use_immediate_energy_parts:
+                old_max = self._maximum_energy(old_resources)
+                new_max = self._maximum_energy(new_resources)
+
+                result = self._duplicate()
+                result._energy = self._energy + (new_max - old_max)
+                return result
+            if (new_resources[self._energy_part_item] // 4) > (old_resources[self._energy_part_item] // 4):
+                return self._at_maximum_energy(new_resources)
+            return self
+        else:
+            return self
+
+    @override
+    def resource_requirements_for_satisfying_damage(self, damage: int) -> list[list[ResourceRequirement]]:
+        # A requirement for many "Energy Tanks" is added,
+        # which is then decreased by how many tanks is in the state by pickups_to_solve_list
+        # FIXME: get the required items for reductions (aka suits)
+        tank_count = 1 + (damage - self._starting_energy) // self._energy_per_tank
+        part_count = 1 + (damage - self._starting_energy) // (self._energy_per_tank // 4)
+        if not self._use_immediate_energy_parts:
+            part_count += (4 - (part_count % 4)) % 4
```

Comment:
the math in this function is kinda confusing in general, but this line in particular sticks out. is that final outer modulo even doing anything?

Reply:
> Would it be cleaner to remove it and add and part_count % 4 != 0 to the if?

I would prefer that